### PR TITLE
Implemented auto UI Scaling :D

### PR DESCRIPTION
**Pull request by Jezithyr**
- Implemented autoUI scaling, scaling supports down to 520 by 520 pixel screens(Idk why you'd even have this)
- UI scaling can be adjusted by overriding some properties in the window root class

https://user-images.githubusercontent.com/6192499/158044030-f7f94e03-ca3c-4819-95e3-6f6fef6f0075.mp4

### CODE REVIEW

File: Robust.Client/UserInterface/UserInterfaceManager.cs [hunk 1]
```diff
@@ -954,11 +956,41 @@ private void WindowContentScaleChanged(WindowContentScaleEventArgs args)
...
+                scaleRatioX = (scaleRatioX > autoScaleMin) ? scaleRatioX : autoScaleMin;
+
+                scaleRatioY = (windowSize.Y > maxScaleRes.Y) ? osScale : ((float)windowSize.Y-minScaleRes.Y)/(maxScaleRes.Y-minScaleRes.Y)*osScale;
+                scaleRatioY = (scaleRatioY > autoScaleMin) ? scaleRatioY : autoScaleMin;
+
+                Logger.DebugS("DEBG", windowSize.ToString());
+                Logger.DebugS("DEBG", scaleRatioX.ToString());
+                Logger.DebugS("DEBG", scaleRatioY.ToString());
+                //Take the smallest UIScale value and use it for UI scaling
+                return (scaleRatioX < scaleRatioY) ? scaleRatioX : scaleRatioY;
```
Comment (mirrorcult) [inline Robust.Client/UserInterface/UserInterfaceManager.cs]: indentation is fucked here

File: Robust.Client/UserInterface/UserInterfaceManager.cs [hunk 2]
```diff
@@ -954,11 +956,41 @@ private void WindowContentScaleChanged(WindowContentScaleEventArgs args)
...
+
+                scaleRatioX = (windowSize.X > maxScaleRes.X) ? osScale : ((float)windowSize.X-minScaleRes.X)/(maxScaleRes.X-minScaleRes.X)*osScale;
+                scaleRatioX = (scaleRatioX > autoScaleMin) ? scaleRatioX : autoScaleMin;
+
+                scaleRatioY = (windowSize.Y > maxScaleRes.Y) ? osScale : ((float)windowSize.Y-minScaleRes.Y)/(maxScaleRes.Y-minScaleRes.Y)*osScale;
+                scaleRatioY = (scaleRatioY > autoScaleMin) ? scaleRatioY : autoScaleMin;
+
+                Logger.DebugS("DEBG", windowSize.ToString());
+                Logger.DebugS("DEBG", scaleRatioX.ToString());
+                Logger.DebugS("DEBG", scaleRatioY.ToString());
```
Comment (mirrorcult) [inline Robust.Client/UserInterface/UserInterfaceManager.cs]: leftover

**Comment by mirrorcult** [comment]
@PJB3005

File: Robust.Client/UserInterface/UserInterfaceManager.cs [hunk 3]
```diff
@@ -954,11 +956,33 @@ private void WindowContentScaleChanged(WindowContentScaleEventArgs args)
...
+            if (!root.AutoScale && (windowSize.X <= 0 || windowSize.Y <= 0)) return osScale;
+            var maxScaleRes = root.AutoScaleUpperCutoff;
+            var minScaleRes = root.AutoScaleLowerCutoff;
+            var autoScaleMin = root.AutoScaleMinimum;
+            float scaleRatioX;
+            float scaleRatioY;
+            
+            //Calculate the scale ratios and clamp it between the maximums and minimums
+            scaleRatioX = (windowSize.X > maxScaleRes.X) ? osScale : ((float)windowSize.X-minScaleRes.X)/(maxScaleRes.X-minScaleRes.X)*osScale;
+            scaleRatioX = (scaleRatioX > autoScaleMin) ? scaleRatioX : autoScaleMin;
```
Comment (PaulRitter) [inline Robust.Client/UserInterface/UserInterfaceManager.cs]: i think you can condense this into
```suggestion
            scaleRatioX = Math.Clamp(((float)windowSize.X-minScaleRes.X)/(maxScaleRes.X-minScaleRes.X)*osScale, autoScaleMin, osScale)
```

File: Robust.Client/UserInterface/UserInterfaceManager.cs [hunk 4]
```diff
@@ -954,11 +956,33 @@ private void WindowContentScaleChanged(WindowContentScaleEventArgs args)
...
+
+            scaleRatioY = (windowSize.Y > maxScaleRes.Y) ? osScale : ((float)windowSize.Y-minScaleRes.Y)/(maxScaleRes.Y-minScaleRes.Y)*osScale;
+            scaleRatioY = (scaleRatioY > autoScaleMin) ? scaleRatioY : autoScaleMin;
```
Comment (PaulRitter) [inline Robust.Client/UserInterface/UserInterfaceManager.cs]: same here

File: Robust.Client/UserInterface/UserInterfaceManager.cs [hunk 5]
```diff
@@ -954,11 +956,33 @@ private void WindowContentScaleChanged(WindowContentScaleEventArgs args)
...
+            //Take the smallest UIScale value and use it for UI scaling
+            return (scaleRatioX < scaleRatioY) ? scaleRatioX : scaleRatioY;
```
Comment (PaulRitter) [inline Robust.Client/UserInterface/UserInterfaceManager.cs]:
```suggestion
            return Math.Min(scaleRatioX, scaleRatioY);
```

**Review by PaulRitter — CHANGES_REQUESTED** [review]

File: Robust.Client/UserInterface/Control.Layout.cs [hunk 6]
```diff
@@ -73,6 +73,29 @@ protected internal virtual void UIScaleChanged()
...
+        ///     Maximum resolution to start clamping autos scale to autoscale minimum
+        /// </summary>
+        [ViewVariables]
+        public virtual Vector2i AutoScaleLowerCutoff => new Vector2i(520, 520);
+
+        /// <summary>
+        ///     The minimum ui scale value that autoscale will scale to
+        /// </summary>
+        [ViewVariables] public virtual float AutoScaleMinimum => 0.5f;
+
```
Comment (PJB3005) [inline Robust.Client/UserInterface/Control.Layout.cs]: These properties shouldn't be on `Control` if they only get used by `WindowRoot`.

File: Robust.Client/UserInterface/UserInterfaceManager.cs (hunk 4 above)
Comment (Jezithyr) [reply Robust.Client/UserInterface/UserInterfaceManager.cs]: Fixed

File: Robust.Client/UserInterface/Control.Layout.cs [hunk 7]
```diff
@@ -73,6 +73,29 @@ protected internal virtual void UIScaleChanged()
         [ViewVariables]
         public virtual float UIScale => Root?.UIScale ?? 1;
 
+        /// <summary>
+        ///     Enable the UI autoscale system, this will scale down the UI for lower resolutions
+        /// </summary>
+        [ViewVariables]
+        public virtual bool AutoScale => false;
```
Comment (PJB3005) [inline Robust.Client/UserInterface/Control.Layout.cs]: All of these are get-only computer properties so you can't actually change them. You probably want `{ get; set; } = initialValue;` instead.

File: Robust.Client/UserInterface/UserInterfaceManager.cs [hunk 8]
```diff
@@ -954,11 +956,32 @@ private void WindowContentScaleChanged(WindowContentScaleEventArgs args)
...
+            var minScaleRes = root.AutoScaleLowerCutoff;
+            var autoScaleMin = root.AutoScaleMinimum;
+            float scaleRatioX;
+            float scaleRatioY;
+            
+            //Calculate the scale ratios and clamp it between the maximums and minimums
+            scaleRatioX = Math.Clamp(((float)windowSize.X-minScaleRes.X)/(maxScaleRes.X-minScaleRes.X)*osScale, autoScaleMin, osScale)
+
+            scaleRatioY = Math.Clamp(((float)windowSize.Y-minScaleRes.Y)/(maxScaleRes.Y-minScaleRes.Y)*osScale, autoScaleMin, osScale)
+            scaleRatioY = (scaleRatioY > autoScaleMin) ? scaleRatioY : autoScaleMin;
```
Comment (PJB3005) [inline Robust.Client/UserInterface/UserInterfaceManager.cs]: Isn't this already handled by the clamp above?

File: Robust.Client/UserInterface/Control.Layout.cs (hunk 6 above)
Comment (Jezithyr) [reply Robust.Client/UserInterface/Control.Layout.cs]: Fixed by moving to Window root

File: Robust.Client/UserInterface/Control.Layout.cs (hunk 7 above)
Comment (Jezithyr) [reply Robust.Client/UserInterface/Control.Layout.cs]: Fixed

File: Robust.Client/UserInterface/UserInterfaceManager.cs (hunk 8 above)
Comment (Jezithyr) [reply Robust.Client/UserInterface/UserInterfaceManager.cs]: Good catch, yeah I got confused thinking that it was the old code for the clamp below

**Review by Jezithyr — COMMENTED** [review]
Fixed

**Pull request merged**